### PR TITLE
Move a bunch of functions from `Type` to `TypePtr`

### PR DESCRIPTION
**Pull request by jvilk-stripe**
Move a bunch of functions from `Type` to `TypePtr`.

In the next PR, I will move the rest and may introduce some preprocessor/templatey goodness to clean up those case statements.

### Motivation

Required to delete `Type` class and make all `Type` objects smaller.

### Test plan

Covered by existing tests.

### CODE REVIEW

File: core/TypePtr.cc [hunk 1]
```diff
@@ -91,4 +91,325 @@ bool TypePtr::isBottom() const {
...
+            return 2;
+        case Tag::LiteralType:
+            return 3;
+        case Tag::ShapeType:
+            return 4;
+        case Tag::TupleType:
+            return 5;
+        case Tag::LambdaParam:
+        case Tag::SelfTypeParam:
+            return 6;
```
Comment (elliottt) [inline core/TypePtr.cc:111]: It's unfortunate how close `kind()` is to `tag()`.

File: core/TypePtr.cc [hunk 2]
```diff
@@ -91,4 +91,325 @@ bool TypePtr::isBottom() const {
...
+            return "OrType";
+        case Tag::AndType:
+            return "AndType";
+        case Tag::SelfType:
+            return "SelfType";
+    }
+}
+
+bool TypePtr::isFullyDefined() const {
+    switch (tag()) {
```
Comment (elliottt) [inline core/TypePtr.cc:165]: This version of `isFullyDefined` is really easy to reason about, with most of the cases all in one place 👍

File: core/types/calls.cc [hunk 3]
```diff
@@ -1103,25 +1099,28 @@ TypePtr getMethodArguments(const GlobalState &gs, SymbolRef klass, NameRef name,
     return TupleType::build(gs, move(args));
 }
 
-TypePtr ClassType::getCallArguments(const GlobalState &gs, NameRef name) {
+TypePtr ClassType::getCallArguments(const GlobalState &gs, NameRef name) const {
     if (hasUntyped()) {
-        return Types::untyped(gs, untypedBlame());
+        return Types::untyped(gs, Symbols::noSymbol());
```
Comment (elliottt) [inline core/types/calls.cc:1104]: Nice: it's OK to use `Symbols::noSymbol()` here because of the specialized implementation for `BlameUntyped` below.

**Review by elliottt — APPROVED** [review]
This change is great! I'm definitely biased, but I much prefer pattern-matching (or matching on tags in this case) to overloads 👍

File: core/TypePtr.cc (hunk 1 above)
Comment (jvilk-stripe) [reply core/TypePtr.cc:111]: If there weren't ties I'd just align the enum type to match `kind`. But there are ties, unfortunately.

**Pull request merged**